### ISSUE TITLE
When running OMV 6.x and you need to apply configuration changes you don't know what they are

### ISSUE THREAD

**Issue by robvdl** [openmediavault/openmediavault#1377]
When you run OMV 6.x on Debian and you receive an update to the openmediavault package, sometimes I will get an email that configuration changes need to be applied.

I then go into the interface and apply them, but you don't really know what you're applying.

Is there no way this can't happen as part of the package update? Going into the interface and applying what you don't know seems a bit strange.

**Comment by votdev** [comment]
I'm sorry to say, but it is not possible to convert the list of dirty modules into a human readable and understandable representation at the current state of the project and the amount of available plugins. It does not make sense to introduce an interface that is only used by a minimal number of plugins. I DON NOT want to implement something that is half-baked.

**Comment by robvdl** [comment]
Fair enough, which is why I was asking if it is not possible to just apply this during package upgrade, might be less confusing to the end user.

Feel free to close this issue then.

**Comment by votdev** [comment]
> if it is not possible to just apply this during package upgrade

This was done before OMV5, but IMO this is a bad approach and because of that it has been changed to the current workflow. I do not want to decide when the configuration is applied and deployed. The admin is responsible for this.

**Comment by sebeksd** [comment]
> This was done before OMV5, but IMO this is a bad approach and because of that it has been changed to the current workflow. I do not want to decide when the configuration is applied and deployed. The admin is responsible for this to decide this.

That is true but in the same time if you don't know what this changes are than you as an admin "Apply" this changes automatically any way without any conscious thought. So till there is no way to review what changes are waiting then there is no reason to not to apply them automatically. One situation when apply should not be automatic is when there was some changes to apply before the update.

**Comment by votdev** [comment]
> > This was done before OMV5, but IMO this is a bad approach and because of that it has been changed to the current workflow. I do not want to decide when the configuration is applied and deployed. The admin is responsible for this to decide this.
> 
> That is true but in the same time if you don't know what this changes are than you as an admin "Apply" this changes automatically any way without any conscious thought. So till there is no way to review what changes are waiting then there is no reason to not to apply them automatically. One situation when apply should not be automatic is when there was some changes to apply before the update.

I'll stop this discussion here because it does not bring any benefit.
There are technical reasons why this `Apply` workflow has been introduced. I have mentioned it 1000 times, but i will repeat myself again: Applying configuration changes immediately does not work out because the deployment of the new configuration might take several seconds to minutes depending on what to change and how powerful the hardware is. It will frustrate the user more than it is necessary. Because of that and to limit the delay of the configuration deployment it is only done when it is triggered by the user. This way several config changes are applied at once which minimizes the delay to a minimum.
I also mentioned 1000 times that there is no way to display the changes in a way a normal user understands; or do you like to see a XML diff? If there would be a way, be sure, i would have implemented it already. I'm not stupid and i know what needs to be done.

**Comment by sebeksd** [comment]
> > > This was done before OMV5, but IMO this is a bad approach and because of that it has been changed to the current workflow. I do not want to decide when the configuration is applied and deployed. The admin is responsible for this to decide this.
> > 
> > 
> > That is true but in the same time if you don't know what this changes are than you as an admin "Apply" this changes automatically any way without any conscious thought. So till there is no way to review what changes are waiting then there is no reason to not to apply them automatically. One situation when apply should not be automatic is when there was some changes to apply before the update.
> 
> I'll stop this discussion here because it does not bring any benefit. There are technical reasons why this `Apply` workflow has been introduced. I have mentioned it 1000 times, but i will repeat myself again: Applying configuration changes immediately does not work out because the deployment of the new configuration might take several seconds to minutes depending on what to change and how powerful the hardware is. It will frustrate the user more than it is necessary. Because of that and to limit the delay of the configuration deployment it is only done when it is triggered by the user. This way several config changes are applied at once which minimizes the delay to a minimum. I also mentioned 1000 times that there is no way to display the changes in a way a normal user understands; or do you like to see a XML diff? If there would be a way, be sure, i would have implemented it already. I'm not stupid and i know what needs to be done.

Could you for example add OMV env for "always apply"? This would be great if it work also from "apt upgrade" (this is my preferred way of doing updates),

**Comment by mi-hol** [comment]
> There are technical reasons why this `Apply` workflow has been introduced. I have mentioned it 1000 times, but i will repeat myself again: Applying configuration changes immediately does not work out because the deployment of the new configuration might take several seconds to minutes depending on what to change and how powerful the hardware is. It will frustrate the user more than it is necessary. Because of that and to limit the delay of the configuration deployment it is only done when it is triggered by the user. This way several config changes are applied at once which minimizes the delay to a minimum.

Unfortunately emotions are involved, could we try to come back to facts?

We understand stand the reasoning for above outlined process too. 

What seems missing for the scenario is the actual use-case.
The current workflow seem to be targeted at 24/7 operation and a skill admin.
Did I get this right?

From my view the users asking for less manual involvement are not skilled admins and accept a downtime outside their normal usage hours.

In summary the requirements are completely opposite.

A solution could be achieved by defining downtimes as "on demand by admin" or "at a certain time in a tbd frequency (i.e. daily"

**Comment by ryecoaaron** [comment]
The api calls wouldn't tell you much since they might exec OS commands and mark modules as dirty. Clicking apply always executes saltstack modules for the dirty modules.  A diff of the database, which is xml, would be horrible since xml doesn't diff well.  Logs of what saltstack is doing might be somewhat useful but wouldn't be everything since some changes are made in php.

**Comment by votdev** [comment]
With https://github.com/openmediavault/openmediavault/pull/1411 a framework has been introduced that can be used to add a brief description about the `dirty` modules. Now it's on the community to contribute these descriptions. I've already included some examples in the PR.

Examples:
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/openmediavault/engined/module/avahi.inc#L34
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/openmediavault/engined/module/cronapt.inc#L30
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/openmediavault/engined/module/monit.inc#L30

The modules of interest can be found here:
- https://github.com/openmediavault/openmediavault/tree/master/deb/openmediavault/usr/share/openmediavault/engined/module
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-clamav/usr/share/openmediavault/engined/module/clamav.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-filebrowser/usr/share/openmediavault/engined/module/filebrowser.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-forkeddaapd/usr/share/openmediavault/engined/module/forkeddaapd.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-nut/usr/share/openmediavault/engined/module/nut.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-onedrive/usr/share/openmediavault/engined/module/onedrive.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-owntone/usr/share/openmediavault/engined/module/owntone.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-photoprism/usr/share/openmediavault/engined/module/photoprism.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-s3/usr/share/openmediavault/engined/module/minio.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-shairport/usr/share/openmediavault/engined/module/shairport.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-snmp/usr/share/openmediavault/engined/module/snmp.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-tftp/usr/share/openmediavault/engined/module/tftp.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-usbbackup/usr/share/openmediavault/engined/module/usbbackup.inc
- https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault-wetty/usr/share/openmediavault/engined/module/wetty.inc
-
